src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DefaultObjectValidator.cs

```diff
@@ -229,8 +241,10 @@ private bool ValidateElements(string currentKey, IEnumerable model, ValidationCo
             var validatorsAsCollection = validators as ICollection;
             if (validatorsAsCollection == null || validatorsAsCollection.Count > 0)
             {
-                var modelValidationContext =
-                        new ModelValidationContext(validationContext.ModelValidationContext, modelExplorer);
+                var modelValidationContext = ModelValidationContext.GetChildValidationContext(
+                    validationContext.ModelValidationContext,
+                    modelExplorer);
+
                 var modelState = validationContext.ModelValidationContext.ModelState;
```

Comment:
The main fix here is to consider BindingSource for a top level parameter as well. Other stuff is cleaning up so it is more understandable.